### PR TITLE
Fix docs-only CI guard after descendant workflow success

### LINKED ISSUE #5024: Follow-up: docs-safety guard can be permanently blocked by a tree-inherent failure on an ancestor commit
## Summary

`ensure-main-docs-safety` walks back through guard-only-failure commits to find "the underlying CI state", and stops at the first commit with a real failure. When that real failure is **inherent to the commit's tree** — a bad link, a bad file — it can never be repaired, because re-running the workflow re-checks-out the same bad tree. The docs-only CI fast path then stays blocked on every subsequent docs-only push until an unrelated non-docs commit lands green.

We are in that state on `main` right now.

## How it happened

1. CodeRabbit's Strapi CDN stopped terminating TLS, killing two README logo URLs (fixed in #5022).
2. `Check Markdown Links #9461` failed on `c44763524` — a **real** failure, not guard-only.
3. Every later docs-only push failed `detect-changes` (17 of them), each one a guard-only failure.
4. #5022 landed as `f8e0be41` and `Check Markdown Links` is **green** there.

But the guard still blocks the next docs-only push, because:

```
checkSha(f8e0be41)  -> failingRuns = 0, guardOnlyRuns = 9  -> recurse to first parent
checkSha(c4476352)  -> Check Markdown Links #9461 = real failure -> setFailed
```

Verified: every failing run on `f8e0be41` failed in `detect-changes` alone, so all are classified guard-only:

```
Lint JS and Ruby       failed-jobs=[detect-changes]
Playwright E2E Tests   failed-jobs=[detect-changes]
Rspec test for gem     failed-jobs=[detect-changes]
Integration Tests      failed-jobs=[detect-changes]
```

and `Check Markdown Links` on `f8e0be41` is `completed/success`.

## Why re-running does not help

`c44763524`'s tree still contains the dead Strapi URLs. Re-running `Check Markdown Links #9461` checks out that tree and fails again, by construction. The fix exists only in a *descendant* commit, but the walk only ever moves toward ancestors.

So the barrier is permanent for the ancestor path. The only documented escape is the guard's own advice — "push non-docs changes to trigger hosted CI" — after which the *next* docs-only push sees a commit with `failingRuns = 0, guardOnlyRuns = 0` and passes.

## The general defect

The walk answers "was CI healthy before this docs-only run?" by looking backwards. That is right for *transient* failures but wrong for *tree-inherent* ones, where the newest commit is strictly better evidence than its ancestors. A single tree-inherent failure on `main` converts into an indefinite repo-wide block on docs-only CI skipping.

This has extra bite because the triggering failure can be entirely external — here, a third-party CDN's TLS config. A vendor decommissioning an asset should not gate this repo's CI routing.

## Proposed fixes

Not mutually exclusive; (1) is the smallest real fix.

1. **Stop the walk at a green run for the same workflow.** Before recursing, check whether the workflow that failed on the ancestor has since concluded successfully on a descendant (or on the checked commit itself). Here, `Check Markdown Links` is green on `f8e0be41`, which should retire the `#9461` failure rather than the walk treating it as live.
2. **Bound staleness by time or distance.** A real failure more than N commits or H hours behind the pushed commit should warn rather than block; the newer green signal is better evidence.
3. **Exclude pure external-link failures from the guard.** `Check Markdown Links` measures third-party uptime as much as repo health. #4989 already made this argument for PR-time noise; this issue shows the same check also gating CI *routing*, which is a larger blast radius. `EXCLUDE_WORKFLOWS` already exists (currently `Benchmark Workflow`) and may be the one-line mitigation.

## Repro / references

- Guard: `.github/actions/ensure-main-docs-safety/check-previous-main.cjs` (walk at ~L464-505, `MAX_GUARD_ONLY_HOPS = 10`)
- Blocked run: https://github.com/shakacode/react_on_rails/actions/runs/34027592631 (`Generator tests` on `f8e0be41`)
- Root-cause failure: https://github.com/shakacode/react_on_rails/actions/runs/34023643033 (`Check Markdown Links #9461` on `c44763524`)
- Fix PR for the underlying links: #5022
- Related tooling issue: #5012

### PR DESCRIPTION
**Pull request by justin808**
## Why

The docs-only CI safety guard walks first-parent history past guard-only failures. A real workflow failure on an older tree could remain a permanent blocker even after the same workflow and failed job passed on a descendant that fixed the defect.

Fixes #5024.

## What changed

- Carry successful workflow-and-job evidence per individual descendant run while walking toward older first parents.
- Match a newer success to an older failure only by immutable numeric `workflow_id` and the exact failed job display name.
- Require a completed trusted run and at least one successful quality-gate job.
- Exclude CI-routing jobs (`detect-changes`, `setup-matrix`, and `setup-integration-matrix`) from success evidence.
- Reject ambiguous duplicate job display names rather than allowing one successful job to hide another skipped or failing job.
- Fail closed before grouping when the API returns a missing or malformed `workflow_id`.
- Keep unrelated failures, guard-only runs, incomplete runs, empty jobs, all-skipped jobs, manual runs, and traversal-limit failures blocking as before.

## How to review and verify

Review the success-evidence predicate in `.github/actions/ensure-main-docs-safety/check-previous-main.cjs`. Then review the same-workflow, cross-job, duplicate-name, and negative regressions in `check-previous-main.test.cjs`.

### Commands and results

- `node .github/actions/ensure-main-docs-safety/check-previous-main.test.cjs` — passed.
- `node --check` for both changed `.cjs` files — passed.
- targeted ESLint and Prettier checks for the changed action files — passed.
- `pnpm run lint` — passed.
- `actionlint` — passed.
- mandatory OSS RuboCop — 252 files inspected, no offenses.
- `git diff --check origin/main...HEAD` — passed.
- `.agents/bin/validate --changed` — all changed-surface checks passed; its expanded local full suite reached 43/45 suites and 665/673 tests before two unrelated Redis-dependent Pro suites failed because local Redis was unavailable.

### Exact-head and replay evidence

- Base: `7cce8e068e43e91e544866254ed2bb06b9b221d7`
- Head: `c40dd303ea2ca17b19ffc0fc638c1be0aa42ad48`
- Red replay: a successful job in a narrower descendant matrix previously suppressed a different skipped/failing ancestor job; a duplicate successful job name could hide a same-named skipped job; and partial successes from two descendant runs could be pooled to hide a combined ancestor failure.
- Green replay: all three regressions now remain blocking, while the intended same-workflow, same-unambiguous-job success from one actual descendant run retires the older failure.

### Workflow Change Audit

- Secret references: unchanged.
- `permissions:`: unchanged.
- `on:` triggers: unchanged.
- Third-party actions: none added or changed.
- Behavior: the local composite action now retires an older failure only when one trusted descendant run completed successfully and the same workflow ID plus every exact, unambiguous failed job name succeeded together in that run.
- Fail-closed controls: invalid IDs and ambiguous job names cannot suppress failures; non-quality evidence, unrelated failures, and traversal exhaustion remain blocking.
- Post-merge exercise: #5054 tracks the required real `push` exercise from a later valuable docs-only PR.

### Review

Independent read-only review of the current repair returned clean after verifying job-specific matching, duplicate-name fail-closed behavior, and per-run evidence grouping. The current-head hosted review is still in progress.

---

> [!NOTE]
> **Medium Risk**
> Changes main-branch docs-only CI gating logic with nuanced job/workflow matching; incorrect supersession could allow skips when CI is still red, though new fail-closed rules and tests target that risk.
> 
> **Overview**
> Updates the **ensure-main-docs-safety** composite action so docs-only skips are not blocked forever by an ancestor workflow failure that was already fixed on a newer commit.
> 
> While walking first-parent history, the guard now **records per-descendant-run success evidence** (numeric `workflow_id` plus unambiguous successful quality-gate job names) and **retires matching older failures** only when every failed job name succeeded in the **same** descendant run. Orchestration jobs (`detect-changes`, `setup-matrix`, `setup-integration-matrix`) no longer count as proof of green CI; empty/all-skipped runs, duplicate job display names, partial successes split across runs, and invalid/missing `workflow_id` values **fail closed** and keep blocking.
> 
> **Tests** in `check-previous-main.test.cjs` add broad regression coverage for supersession, matrix/skipped-job edge cases, and API shape validation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3cb2604a07732f3a6bed520a1bd6cc285a344069. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of documentation-only updates when an earlier failed check is superseded by a successful follow-up run.
  * Added safeguards so incomplete, ambiguous, invalid, or jobless workflow results are not treated as successful.
  * Improved validation and evaluation of workflow checks, including complex job and matrix scenarios.

* **Tests**
  * Expanded automated coverage for superseded checks, partial successes, skipped jobs, invalid workflow identifiers, and in-progress runs.

### CODE REVIEW

**Comment by justin808** [comment]
Workflow Change Audit:

- Exact head: `113f2d012d14d2dc082c6f4d06e38aa8c0bfe146`
- Secret references: unchanged.
- `permissions:`: unchanged.
- `on:` triggers: unchanged.
- Third-party actions: none added or version-changed.
- Behavior: the local composite action can retire an older workflow failure only after a trusted descendant run with the same positive safe integer `workflow_id` completes successfully and contains a successful quality-gate job.
- Fail-closed controls: missing or malformed workflow IDs reject the API response; routing-only, guard-only, empty, skipped, incomplete, manual, and unrelated workflow evidence cannot suppress an older failure; traversal limits remain unchanged.
- Post-merge exercise: #5054.
- Focused action suite, full JS lint, full formatting, syntax checks, actionlint, and diff checks passed on this head. Two independent exact-head reviews returned `READY` with no open findings.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 1]
```diff
@@ -16,10 +17,22 @@ function summarizeRun(run) {
...
 
+function isValidWorkflowId(workflowId) {
+  return Number.isSafeInteger(workflowId) && workflowId > 0;
+}
+
 function latestRunsByWorkflow(workflowRuns) {
   const latestByWorkflow = new Map();
 
   for (const run of workflowRuns) {
+    if (!isValidWorkflowId(run.workflow_id)) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:28]: [auto-deferred]

The ordering is intentionally fail-closed: every trusted API workflow record is validated before exclusions. GitHub supplies numeric workflow IDs, and moving exclusions earlier would weaken malformed-response detection without fixing a live failure. Retaining this defensive behavior avoids final-candidate nit churn.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 2]
```diff
@@ -235,7 +252,14 @@ async function evaluateCommitRuns({ github, context, core, sha, createdAfter, ex
       const failed = failedJobs(latestJobs);
 
       if (failed.length === 0) {
-        return { kind: 'passing' };
+        const hasSuccessfulSubstantiveJob = latestJobs.some(isSuccessfulQualityGateJob);
+        return {
+          kind: 'passing',
+          successfulWorkflowId:
+            run.conclusion === 'success' && isValidWorkflowId(run.workflow_id) && hasSuccessfulSubstantiveJob
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs]: [auto-deferred]

Confirmed as harmless defensive redundancy at the trust boundary. Keeping the checks makes the local safety invariant explicit and avoids final-candidate nit-only churn.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 3]
```diff
@@ -235,7 +252,14 @@ async function evaluateCommitRuns({ github, context, core, sha, createdAfter, ex
...
+              ? run.workflow_id
+              : null,
+        };
       }
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:286]: Fixed in 63b6a8f7ab0b771c324ccb96131f47dc13fe7b66. Descendant success evidence is now keyed by workflow ID plus exact failed job display names, and ambiguous duplicate names fail closed. Added red/green regressions for both a different skipped job in the same workflow and duplicate success/skipped display names. The focused helper suite, lint, formatting, syntax checks, actionlint, and independent review all pass.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 4]
```diff
@@ -501,6 +584,13 @@ async function checkPreviousMainCommitStatus({
     core.info(
       `Main commit ${shaToCheck} only has docs-only guard failures. Checking first parent ${parentSha} for the underlying CI state.`,
     );
+    for (const [workflowId, jobNames] of result.successfulJobsByWorkflow) {
+      const accumulatedJobNames = successfulDescendantJobsByWorkflow.get(workflowId) || new Set();
+      for (const jobName of jobNames) {
+        accumulatedJobNames.add(jobName);
+      }
+      successfulDescendantJobsByWorkflow.set(workflowId, accumulatedJobNames);
+    }
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:593]: Fixed in c40dd303ea2ca17b19ffc0fc638c1be0aa42ad48. Successful job evidence now remains grouped as one set per actual descendant workflow run; an ancestor failure is superseded only when one set covers every exact, unambiguous failed job name. Added a red/green two-hop regression proving partial successes from separate runs cannot be pooled. Focused tests, syntax, ESLint, Prettier, full JS lint, actionlint, mandatory OSS RuboCop, and independent Codex review pass.

**Comment by justin808** [comment]
Address-review complete for exact head `c40dd303ea2ca17b19ffc0fc638c1be0aa42ad48` (cutoff `2026-09-11T23:35:05Z`).

- Fixed and resolved: workflow-level success could hide a different failed job; duplicate job names could be ambiguous; successes from separate descendant runs could be pooled.
- Deliberately retained and resolved: fail-closed workflow-ID validation ordering and defensive ID checks.
- Verification: focused action tests, syntax checks, ESLint, Prettier, full JS lint, actionlint, mandatory OSS RuboCop, independent Codex review, and current-head Claude review passed.
- Unresolved review threads: 0.
- Remaining gate: exact-head `high-risk-files` acknowledgement before readiness and one optimized hosted validation cohort.

**Comment by justin808** [comment]
+ci-run-hosted

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 5]
```diff
@@ -233,36 +267,57 @@ async function evaluateCommitRuns({ github, context, core, sha, createdAfter, ex
...
   const guardOnlyRuns = [];
+  const successfulJobsByWorkflow = new Map();
 
   for (const runResult of completedRunResults) {
     if (runResult.kind === 'failing') {
-      failingRuns.push(runResult.run);
+      failingRunResults.push(runResult);
     } else if (runResult.kind === 'guard-only') {
       guardOnlyRuns.push(runResult.run);
+    } else if (runResult.successfulWorkflowId !== null) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs]: Fixed in cd8374862157df28594a61f48ac55aff99ce0d69. Successful workflow evidence is now recorded only for `kind === passing`. A direct regression verifies that an empty workflow run cannot create an `undefined` success-evidence entry. Focused tests, syntax checks, ESLint, Prettier, actionlint, and mandatory OSS RuboCop pass.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 6]
```diff
@@ -461,8 +517,33 @@ async function checkPreviousMainCommitStatus({
...
+          ),
+        )
+      );
+    };
+    const supersededFailingRuns = result.failingRunResults
+      .filter(isSupersededByDescendant)
+      .map(({ run }) => run);
+    const unresolvedFailingRuns = result.failingRunResults
+      .filter((runResult) => !isSupersededByDescendant(runResult))
+      .map(({ run }) => run);
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:543]: [auto-deferred]

Declined at the final-candidate debounce point. The predicate is pure, the list is small, and the duplicate evaluation has no correctness or meaningful performance cost. Refactoring would restart exact-head review and CI without material benefit.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 7]
```diff
@@ -16,10 +17,22 @@ function summarizeRun(run) {
...
   const latestByWorkflow = new Map();
 
   for (const run of workflowRuns) {
+    if (!isValidWorkflowId(run.workflow_id)) {
+      const error = new TypeError(
+        `Expected workflow run ${run.id ?? 'UNKNOWN'} to have a positive safe integer workflow_id.`,
+      );
+      error.unexpectedGithubApiResponse = true;
+      throw error;
+    }
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:34]: [auto-deferred]

Duplicate of the resolved fail-closed decision in discussion_r3988257097. The action intentionally validates every trusted workflow record before exclusions, so malformed GitHub API data cannot bypass the guard. No code change.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs (hunk 6 above)
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:543]: [auto-deferred]

Duplicate of the resolved optional suggestion in discussion_r3994346377. The predicate is pure, the list is small, and duplicate evaluation has no correctness or meaningful performance cost. No final-candidate refactor.
